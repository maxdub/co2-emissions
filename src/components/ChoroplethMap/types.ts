type Co2Data = {
    co2: number;
    co2_per_capita: number;
    coal_co2: number;
    coal_co2_per_capita: number;
    cumulative_co2: number;
    cumulative_coal_co2: number;
    population: number;
    share_global_co2: number;
    share_global_coal_co2: number;
    share_global_cumulative_co2: number;
    share_global_cumulative_coal_co2: number;
    year: number;
};

type CountryData = {
    iso_code: string;
    data: Array<Co2Data>;
};

export type CountryCo2Data = {
    [key: string]: CountryData;
};
