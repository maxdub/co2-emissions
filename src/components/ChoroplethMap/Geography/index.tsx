import React from "react";

function Geography() {
    return <></>;
}

export default Geography;
