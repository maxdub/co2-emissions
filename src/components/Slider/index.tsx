import React from "react";
import MaterialSlider from "@material-ui/core/slider";
import styled from "styled-components/macro";

const marks = [
    {
        value: 1864,
        label: "1864",
    },
    {
        value: 1900,
        label: "1900",
    },
    {
        value: 1950,
        label: "1950",
    },
    {
        value: 2000,
        label: "2000",
    },
    {
        value: 2019,
        label: "2019",
    },
];

export type Props = {
    curYear: number;
    onChange: (event: React.ChangeEvent<{}>, value: number | number[]) => void;
};

function Slider({ curYear, onChange }: Props) {
    return (
        <SliderContainer>
            <SliderLabel data-testid="slider-label">CO2 Emissions in Year {curYear}</SliderLabel>
            <MaterialSlider
                onChangeCommitted={onChange}
                marks={marks}
                defaultValue={2019}
                min={1864}
                step={1}
                max={2019}
            />
        </SliderContainer>
    );
}

const SliderLabel = styled.p``;

const SliderContainer = styled.div`
    width: 400px;
`;

export default Slider;
