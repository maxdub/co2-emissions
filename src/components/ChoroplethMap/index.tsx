import React, { useEffect, useState } from "react";
import { ComposableMap, Geographies, Sphere, Graticule } from "react-simple-maps";
import { CountryCo2Data } from "./types";
import Geography from "../Geography";
import styled from "styled-components/macro";
import Slider from "../Slider";
import ReactTooltip from "react-tooltip";

const projectionConfig = {
    scale: 147,
};

function ChoroplethMap() {
    const [data, setData] = useState<CountryCo2Data>({});
    const [geo, setGeo] = useState({});
    const [loading, setLoading] = useState(false);
    const [curYear, setYear] = useState<number>(2019);
    const [content, setTooltip] = useState("");

    useEffect(() => {
        setLoading(true);
        Promise.all([
            fetch("https://raw.githubusercontent.com/owid/co2-data/master/owid-co2-data.json")
                .then((response) => response.json())
                .then((data) => setData(data)),
            fetch(
                "https://raw.githubusercontent.com/zcreativelabs/react-simple-maps/master/topojson-maps/world-110m.json",
            )
                .then((response) => response.json())
                .then((data) => setGeo(data)),
        ]).finally(() => {
            setLoading(false);
        });
    }, []);

    if (!data) {
        return null;
    }

    return (
        <>
            <ReactTooltip>{content}</ReactTooltip>
            <Slider
                curYear={curYear}
                onChange={(_evt, value) => {
                    if (Array.isArray(value)) return;
                    setYear(value);
                }}
            />
            {loading && <Loading>Loading CO2 data...</Loading>}
            <ComposableMap data-tip="World" height={500} projectionConfig={projectionConfig}>
                <Sphere fill="#F8F8F8" id="choropleth" stroke="#E4E5E6" strokeWidth={0.5} />
                <Graticule stroke="#E4E5E6" strokeWidth={0.5} />
                {Object.keys(data).length > 0 && (
                    <Geographies geography={geo}>
                        {({ geographies }) =>
                            geographies.map((geo) => (
                                <Geography
                                    key={geo.rsmKey}
                                    setTooltip={setTooltip}
                                    data={data}
                                    curYear={curYear}
                                    geo={geo}
                                />
                            ))
                        }
                    </Geographies>
                )}
            </ComposableMap>
        </>
    );
}

const Loading = styled.p`
    position: absolute;
    top: 140px;
`;

export default ChoroplethMap;
