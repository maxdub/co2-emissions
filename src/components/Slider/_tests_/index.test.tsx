import React from "react";
import { createRenderer } from "react-test-renderer/shallow";
import { render, getByLabelText } from "@testing-library/react";
import Slider from "../";
import { CountryCo2Data } from "../../ChoroplethMap/types";
import { Props } from "../";

const renderer = createRenderer();

const renderComponent = ({ curYear, onChange }: Props) => {
    return render(<Slider curYear={curYear} onChange={onChange} />);
};

describe("<Slider/>", () => {
    let output: ReturnType<typeof renderComponent>;
    beforeEach(() => {
        output = renderComponent({ curYear: 1991, onChange: () => null });
    });

    it("should render and match snapshot", () => {
        expect(output.container.firstChild).toMatchSnapshot();
    });

    it("It's current year should be the input year of 1991", () => {
        expect(output.getByTestId("slider-label")).toHaveTextContent("CO2 Emissions in Year 1991");
    });
});
