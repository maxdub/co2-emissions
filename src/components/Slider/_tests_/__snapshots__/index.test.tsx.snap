// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Slider/> should render and match snapshot 1`] = `
<div
  class="Slider__SliderContainer-l2gxtt-1 fkazSp"
>
  <p
    class="Slider__SliderLabel-l2gxtt-0"
    data-testid="slider-label"
  >
    CO2 Emissions in Year 
    1991
  </p>
  <span
    class="MuiSlider-root MuiSlider-colorPrimary MuiSlider-marked"
  >
    <span
      class="MuiSlider-rail"
    />
    <span
      class="MuiSlider-track"
      style="left: 0%; width: 100%;"
    />
    <input
      type="hidden"
      value="2019"
    />
    <span
      class="MuiSlider-mark MuiSlider-markActive"
      data-index="0"
      style="left: 0%;"
    />
    <span
      aria-hidden="true"
      class="MuiSlider-markLabel MuiSlider-markLabelActive"
      data-index="0"
      style="left: 0%;"
    >
      1864
    </span>
    <span
      class="MuiSlider-mark MuiSlider-markActive"
      data-index="1"
      style="left: 23.225806451612904%;"
    />
    <span
      aria-hidden="true"
      class="MuiSlider-markLabel MuiSlider-markLabelActive"
      data-index="1"
      style="left: 23.225806451612904%;"
    >
      1900
    </span>
    <span
      class="MuiSlider-mark MuiSlider-markActive"
      data-index="2"
      style="left: 55.483870967741936%;"
    />
    <span
      aria-hidden="true"
      class="MuiSlider-markLabel MuiSlider-markLabelActive"
      data-index="2"
      style="left: 55.483870967741936%;"
    >
      1950
    </span>
    <span
      class="MuiSlider-mark MuiSlider-markActive"
      data-index="3"
      style="left: 87.74193548387096%;"
    />
    <span
      aria-hidden="true"
      class="MuiSlider-markLabel MuiSlider-markLabelActive"
      data-index="3"
      style="left: 87.74193548387096%;"
    >
      2000
    </span>
    <span
      class="MuiSlider-mark MuiSlider-markActive"
      data-index="4"
      style="left: 100%;"
    />
    <span
      aria-hidden="true"
      class="MuiSlider-markLabel MuiSlider-markLabelActive"
      data-index="4"
      style="left: 100%;"
    >
      2019
    </span>
    <span
      aria-orientation="horizontal"
      aria-valuemax="2019"
      aria-valuemin="1864"
      aria-valuenow="2019"
      class="MuiSlider-thumb MuiSlider-thumbColorPrimary"
      data-index="0"
      role="slider"
      style="left: 100%;"
      tabindex="0"
    />
  </span>
</div>
`;
